### DISCUSSION TITLE
Questions about modelBuild.sh

### DISCUSSION THREAD

**Discussion by DavidBarriere** [CoBrALab/optimized_antsMultivariateTemplateConstruction#71] [Q&A]
Hi Gabriel,
Thank you for sharing the modelBuild.sh script: it does a wonderful job here on my T1w data.
I have a couple of histological images of brain sections acquired on a slice scanner.
My goal here is to normalize brain section images from 6 or 8 animals (20 sections for each brain) to create a section-by-section 2D template image on which I'd like to normalize my atlas.
Once the 2D section template image is released, I'd like to train a DNN to automatically find the template image corresponding to a newest brain section using an in house data set to automatically normalize my atlas on it and collect histological data from the brain section (grey level, number of positive cells, etc.).

To create a first section-by-section 2D template image, I tried to use modelBuild.sh to realign and normalize my brain sections from three animals as a test but it failed...
Data has been acquired onto a png format that I converted as 2D NIFTI files. If the linear steps worked (but not very well actually) the diffeomorphic steps ran but failed.
I saw that Ants have been previously used to create a face template from a face photography data set, and basically it's what I wanna do with my brain section but I failed to use Ants to redo the job.

So do you have an option for me ?
Dataset and template generate by modelBuild.sh is available here : https://filesender.renater.fr/?s=download&token=6cd55c7f-eb29-4851-a48a-b0fbdd27a5e6

Thank you for your help.

David

Thanks

**Comment by gdevenyi** [comment]
> To create a first section-by-section 2D template image, I tried to use modelBuild.sh to realign and normalize my brain sections from three animals as a test but it failed...

Right now, everything is hard-coded to be `-d 3` for 3D images everywhere in the pipeline.

I've never had any 2D data to work with to even attempt to test.

My suggestion would be to attempt to modify the code and change `-d 3` to `-d 2` in the appropriate places, and `ImageMath 3`  and `AverageImages 3` to `2` as well.

If you can get the hacky version working, I can spend some time to implement a full option feature version.

**Comment by DavidBarriere** [comment]
It sounds easy to change the code waiting for your updates.
I also saw ImageSetStatistics 3 , ResampleImage 3, ThresholdImage 3, ExtractRegionFromImageByMask 3 within the code.
Do I have to change 3 by 2 for these functions as well, or it doesn't really matter ?

**Reply by gdevenyi** [reply]
just to be safe, you should do all of them, they may not be in the codepath, but better safe than sorry.

**Reply by DavidBarriere** [reply]
The test is running :+1:

**Comment by DavidBarriere** [comment]
So, here the modified code failed at the end of the first iteration of the rigid step :
[dbarriere@grappe ~]$ squeue 
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
            167101    bigmem modelbui dbarrier PD       0:00      1 (DependencyNeverSatisfied)
            167102    bigmem modelbui dbarrier PD       0:00      1 (Dependency)
            167103    bigmem modelbui dbarrier PD       0:00      1 (Dependency)
[...]

It's seems that the problem is between the first and the second step since it created template and transform files : 
├── rigid
│   ├── 0
│   │   ├── average
│   │   │   ├── nonzero.nii.gz
│   │   │   ├── template.nii.gz
│   │   │   └── template_sharpen.nii.gz
│   │   ├── resample
│   │   │   ├── brebis_83041.nii.gz
│   │   │   ├── brebis_83060.nii.gz
│   │   │   ├── brebis_83157.nii.gz
│   │   │   └── masks
│   │   └── transforms
│   │       ├── brebis_83041_0GenericAffine.mat
│   │       ├── brebis_83041-to-reference_Warped_SubSampled_0GenericAffine.mat
│   │       ├── brebis_83060_0GenericAffine.mat
│   │       ├── brebis_83060-to-reference_Warped_SubSampled_0GenericAffine.mat
│   │       ├── brebis_83157_0GenericAffine.mat
│   │       └── brebis_83157-to-reference_Warped_SubSampled_0GenericAffine.mat
│   ├── 1
│   │   ├── average
│   │   ├── resample
│   │   │   └── masks
│   │   └── transforms

In attached, I send you the log file
[slurm-modelbuild_2023-04-14_16-21-46-UTC_rigid_0_reg-167100.txt](https://github.com/CoBrALab/optimized_antsMultivariateTemplateConstruction/files/11236002/slurm-modelbuild_2023-04-14_16-21-46-UTC_rigid_0_reg-167100.txt)

**Reply by gdevenyi** [reply]
Well, this is pretty nuts:
`Shrink factors (level 1 out of 388):`

Is the NIFTI metadata accurate about the size/scale of the data? Can you share a `PrintHeader` of the files? How big are they in real-space terms? What is the voxel resolution supposed to be?

**Reply by DavidBarriere** [reply]
Here the header :
dbarriere@grappe 00-rawData_2D]$ PrintHeader brebis_83041.nii.gz
 Spacing [0.5, 0.5]
 Origin [14000, 14000]
 Direction 
1 0
0 1

 Size : 28000 28000 

  Image Dimensions   : [28000, 28000]
  Bounding Box       : {[14000 14000], [28000 28000]}
  Voxel Spacing      : [0.5, 0.5]
  Intensity Range    : [0, 255]
  Mean Intensity     : 88.8896
  Direction Cos Mtx. : 
1 0
0 1

  Voxel->RAS x-form  : 
  Image Metadata: 
    ITK_original_direction of unsupported type N3itk6MatrixIdLj2ELj2EEE
    ITK_original_spacing of unsupported type St6vectorIdSaIdEE
    bitpix = 16
    cal_max = 0
    cal_min = 0
    datatype = 4
    dim[0] = 2
    dim[1] = 28000
    dim[2] = 28000
    dim[3] = 1
    dim[4] = 1
    dim[5] = 1
    dim[6] = 1
    dim[7] = 1
    dim_info = 0
    intent_code = 0
    intent_p1 = 0
    intent_p2 = 0
    intent_p3 = 0
    nifti_type = 1
    pixdim[0] = 0
    pixdim[1] = 0.5
    pixdim[2] = 0.5
    pixdim[3] = 1
    pixdim[4] = 1
    pixdim[5] = 1
    pixdim[6] = 1
    pixdim[7] = 1
    qform_code = 1
    qform_code_name = NIFTI_XFORM_SCANNER_ANAT
    qoffset_x = -14000
    qoffset_y = -14000
    qoffset_z = 0
    quatern_b = 0
    quatern_c = 0
    quatern_d = 1
    scl_inter = 127
    scl_slope = 1
    sform_code = 2
    sform_code_name = NIFTI_XFORM_ALIGNED_ANAT
    slice_code = 0
    slice_duration = 0
    slice_end = 0
    slice_start = 0
    srow_x = 0.4998 0.009996 0 -1492.98
    srow_y = -0.009996 0.4998 0 -2099.14
    srow_z = 0 0 -1 0
    toffset = 0
    vox_offset = 352
    xyzt_units = 0
